source/tests/integration/ni_fake_non_ivi_service_tests_endtoend.cpp

```diff
@@ -65,84 +66,120 @@ class NiFakeNonIviServiceTests_EndToEnd : public ::testing::Test {
   {
     return stub_;
   }
+
+  auto start_read_stream(
+    ::grpc::ClientContext& context,
+    int32_t start,
+    int32_t stop)
+  {
+    ReadStreamRequest request;
+    request.set_start(start);
+    request.set_stop(stop);
+    return stub()->ReadStream(&context, request);
+  }
+
+  template <typename TResponse>
+  bool read_stream(
+    ::grpc::ClientReader<TResponse>& reader,
+    size_t count,
+    std::vector<TResponse>* responses = nullptr)
+  {
+    typename TResponse response;
+    for (size_t i = 0; i < count; ++i) {
+      if (!reader.Read(&response)) {
+        return false;
+      }
+      if (responses) {
+        responses->push_back(response);
+      }
+    }
+    return true;
+  }
+
+  size_t get_range_length(int32_t start, int32_t stop)
+  {
+    return static_cast<size_t>(stop - start + 1);
+  }
+
+  std::vector<int32_t> make_range(int32_t start, int32_t stop)
+  {
+    std::vector<int32_t> range;
+    range.reserve(get_range_length(start, stop));
+    for (int32_t i = start; i <= stop; ++i) {
+      range.push_back(i);
+    }
+    return range;
+  }
+
+  std::vector<int32_t> get_values(const std::vector<ReadStreamResponse>& responses) {
+    std::vector<int32_t> values;
+    for (auto& response : responses) {
+      values.push_back(response.value());
+    }
+    return values;
+  }
 };
 
 TEST_F(NiFakeNonIviServiceTests_EndToEnd, ReadStream_ReturnsExpectedSequence)
 {
   ::grpc::ClientContext context;
   const auto START = 0;
   const auto STOP = 100;
-  ReadStreamRequest request;
-  request.set_start(START);
-  request.set_stop(STOP);
-  auto reader = stub()->ReadStream(&context, request);
+  auto reader = start_read_stream(context, START, STOP);
+  auto expected_values = make_range(START, STOP);
 
-  ReadStreamResponse response;
-  for (auto i = START; i <= STOP; ++i) {
-    reader->Read(&response);
-    EXPECT_EQ(response.value(), i);
-  }
+  std::vector<ReadStreamResponse> responses;
+  bool read_success = read_stream(*reader, expected_values.size(), &responses);
+
+  EXPECT_TRUE(read_success);
+  EXPECT_THAT(get_values(responses), ContainerEq(expected_values));
 }
 
 TEST_F(NiFakeNonIviServiceTests_EndToEnd, ReadPartialStream_ReturnsExpectedSequence)
 {
   ::grpc::ClientContext context;
   const auto START = 0;
   const auto STOP = 100;
-  ReadStreamRequest request;
-  request.set_start(START);
-  request.set_stop(STOP);
-  auto reader = stub()->ReadStream(&context, request);
-
-  ReadStreamResponse response;
-  for (auto i = START; i <= STOP / 2; ++i) {
-    reader->Read(&response);
-    EXPECT_EQ(response.value(), i);
-  }
+  auto reader = start_read_stream(context, START, STOP);
+  auto expected_values = make_range(START, STOP / 2);
+
+  std::vector<ReadStreamResponse> responses;
+  bool read_success = read_stream(*reader, expected_values.size(), &responses);
+
+  EXPECT_TRUE(read_success);
+  EXPECT_THAT(get_values(responses), ContainerEq(expected_values));
```

Comment:
Here is an example of `ContainerEq` assertion output, which was due to an off-by-one error in `make_range`:
```
[ RUN      ] NiFakeNonIviServiceTests_EndToEnd.ReadTwoStreamsInterleaved_ReturnsExpectedSequences
C:\Dev\grpc-device\source\tests\integration\ni_fake_non_ivi_service_tests_endtoend.cpp(220): error: Value of: get_values(responses1)
Expected: equals { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, ... }
  Actual: { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, ... }, which has these unexpected elements: 100
C:\Dev\grpc-device\source\tests\integration\ni_fake_non_ivi_service_tests_endtoend.cpp(221): error: Value of: get_values(responses2)
Expected: equals { 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, ... }
  Actual: { 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, ... }, which has these unexpected elements: 200
[  FAILED  ] NiFakeNonIviServiceTests_EndToEnd.ReadTwoStreamsInterleaved_ReturnsExpectedSequences (36 ms)
```